### PR TITLE
Case Sensitive Amend Badges

### PR DESCRIPTION
**Pull request by dimak1**
*Description of changes:*
- Add case sensitive check for Amend Badges for Transport Permit

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the PPR license (Apache 2.0).

### CODE REVIEW

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-assets-dev--pr-1766-zecotmr4.web.app

File: ppr-ui/src/utils/utilities.ts
Comment (cameron-eyds) [inline ppr-ui/src/utils/utilities.ts]: Why expand the simplified statement into long form with an `if` condition and double returns? 

`return keys1.length !== keys2.length || keys1.some(key => deepChangesComparison(base[key], current[key], isCaseSensitive))` Is the exact same with less code and more concise no?

**Review by cameron-eyds — APPROVED** [review]
Looks good to me! 
Just one question

Edit* I guess another question and maybe a business question, why are we case-sensitive on amendments and not corrections, what was the rationale there from the requirements end?

File: ppr-ui/src/utils/utilities.ts
Comment (dimak1) [reply ppr-ui/src/utils/utilities.ts]: I was getting `...reaching memory limit: JS heap out of memory` error when running the tests. This fixed it, because it does not need to evaluate the second condition (and won't go deeper into recursion).

Screenshot 2024-03-04 at 16 49 42

**Comment by dimak1** [comment]
> Edit* I guess another question and maybe a business question, why are we case-sensitive on amendments and not corrections, what was the rationale there from the requirements end?

@cameron-eyds I asked Liz if we want to make it consistent (twice). Didn't get a clear answer as to why, but she asked to implement the requirement. 🤷🏻

File: ppr-ui/src/utils/utilities.ts
Comment (cameron-eyds) [reply ppr-ui/src/utils/utilities.ts]: This doesn't seem right then, there was no issue before. 

Please look a bit deeper into this, changing code syntax to pass unit tests is somewhat of a red flag, especially considering the simplicity of the statement.

**Review by cameron-eyds — CHANGES_REQUESTED** [review]
Requesting a deeper look into the condition refactored to pass unit tests.

**Comment by cameron-eyds** [comment]
> > Edit* I guess another question and maybe a business question, why are we case-sensitive on amendments and not corrections, what was the rationale there from the requirements end?
> 
> @cameron-eyds I asked Liz if we want to make it consistent (twice). Didn't get a clear answer as to why, but she asked to implement the requirement. 🤷🏻

Fair enough, thank you for inquiring!

**Review by doug-lovett — APPROVED** [review]
Nothing to add to Cameron's review.

File: ppr-ui/src/utils/utilities.ts
Comment (cameron-eyds) [reply ppr-ui/src/utils/utilities.ts]: Let me know if you want me to run the code and help with the unit test debugging.

Comment (dimak1) [reply ppr-ui/src/utils/utilities.ts]: Yea, thanks, I spent few hours today trying to figure it out. I was surprised as well to get this error. But I have another solution, so will try it. Will let you know.

Comment (dimak1) [reply ppr-ui/src/utils/utilities.ts]: Also, the build tests didn't have the same js heap issue.

Comment (cameron-eyds) [reply ppr-ui/src/utils/utilities.ts]: Ok, fair enough, if you've already spent time looking into it than all good. 
It does seem related to test set up, which is why i lean to leaving the code as it was and refactoring the test setup but i'll leave it to you to decide.

**Review by cameron-eyds — APPROVED** [review]

File: ppr-ui/src/utils/utilities.ts
Comment (dimak1) [reply ppr-ui/src/utils/utilities.ts]: I restarted my machine, cleared cache, reinstalled the node, etc. and the tests passed, once... Now it is failing again. Very strange.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-assets-dev--pr-1766-zecotmr4.web.app

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-assets-dev--pr-1766-zecotmr4.web.app

**Comment by dimak1** [comment]
@cameron-eyds yay it's 🟢 please take a look at latest changes, thanks
Edit: tests are passing now, hopefully won't cause issues in future PR.

File: ppr-ui/src/utils/utilities.ts
Comment (dimak1) [reply ppr-ui/src/utils/utilities.ts]: Regarding this, I tried to revert to original logic, but it failed with memory error.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-assets-dev--pr-1766-zecotmr4.web.app

File: ppr-ui/src/utils/utilities.ts
Comment (cameron-eyds) [reply ppr-ui/src/utils/utilities.ts]: Hmm yea, as you've noted in your other PR it's not really related to this specific logic, so i think it's safe to revert and continue to debug, or rerun the unit test suite until it passes since it's intermittent. 

We can focus on this as a single issue and not let it block this PR but we can now be pretty certain it's not related to the || statement, as your now seeing this in all the PR's. 
My recommendation is to revert and continue to run the tests until they pass and we can tackle this issue standalone, i don't think it's related to any code in this PR.

Comment (dimak1) [reply ppr-ui/src/utils/utilities.ts]: Yea I agree it's not probably related to this. However, I still think that we could avoid logical OR, just because we don't want to run the second conditional statement (`keys1.some(key => deepChangesComparison...`) if first condition fails. It might be fast compare, but unnecessary?

Comment (dimak1) [reply ppr-ui/src/utils/utilities.ts]: Updated as per your recommendation.

**Pull request merged**
